kubernetes/node-problem-detector
# node-problem-detector: report disk queue length in Prometheus format

This change is to address below items from #284 :

- [x] Refactor problem daemon registration and initialization, so that they are more modular, and rely on less per-daemon initialization code duplication.
- [x] Refactor events/condition reporting into k8s_exporter.
- [x] Add a SystemStatsMonitor problem daemon, to monitor some system metrics; and add an optional Prometheus metrics exporter.

Since this PR will need OpenCensus and gopsutil library, I submitted PR #289 for it.

Deprecated (but not removed yet) flags in this PR:

- --custom-plugin-monitors
- --system-log-monitors

New flags introduced in this PR:

- --config.custom-plugin-monitor
- --config.system-log-monitor
- --config.system-stats-monitor
- --enable-k8s-exporter
- --prometheus-address
- --prometheus-port

Test steps:

1. Verifying existing command works:
> make ./bin/node-problem-detector && ./bin/node-problem-detector --system-log-monitors=config/kernel-monitor.json --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --logtostderr

2. Testing new flags for existing behavior:
> make ./bin/node-problem-detector && ./bin/node-problem-detector --config.system-log-monitor=config/kernel-monitor.json --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --logtostderr

3. Test NPD for disk monitor with no k8s_exporter
> make ./bin/node-problem-detector && ./bin/node-problem-detector --config.system-log-monitor=config/kernel-monitor.json --apiserver-override=http://127.0.0.1:8080?inClusterConfig=false --logtostderr --config.system-stats-monitor=config/system-stats-monitor.json --enable-k8s-exporter=false

> curl http://localhost:20257/metrics

Follow up:
I will later update test infra to use these new flags.
